### PR TITLE
fence_cdu: new fence agent

### PR DESCRIPTION
**Pull request by watologo1**
This fence has been found in a really old Ubuntu fence-agents version
(distro precise from 2012) and was added by a separate tarball.
It has been re-worked to be compatible with latest fence-agents mainline
package, tested against a SENTRY power switch with firmware version 7.1a.

The original code has been written by Andres Rodriguez. Beside adding myself
with a recent copyright notice reflecting my changes, the original license
description on top of the sources (GPL) is untouched.

No idea why this never made it mainline, but I can confirm it convenient
and working with the HW I tested against.

Main credits for the original fence development go to:

CC: Andres Rodriguez <andres.rodriguez@canonical.com>

### CODE REVIEW

File: agents/cdu/fence_cdu.py [hunk 1]
```diff
@@ -0,0 +1,143 @@
...
+                    outlets[res.group(2)] = (res.group(1))
+            conn.send(action + " " + outlets[options["--plug"]] + "\r\n")
+            conn.log_expect(options["--command-prompt"], int(options["--shell-timeout"]))
+        except pexpect.EOF:
+            fail(EC_CONNECTION_LOST)
+        except pexpect.TIMEOUT:
+            fail(EC_TIMED_OUT)
+
+def main():
+    device_opt = [  "help", "version", "agent", "quiet", "verbose", "debug",
```
Comment (oalbrigt) [inline agents/cdu/fence_cdu.py:86]: You should be able to remove lots of these (some are included by default, and others might be included by e.g. secure or similar (like inetX_only).

You can run `make xml-check` to see if you removed too many :)

File: agents/cdu/fence_cdu.py [hunk 2]
```diff
@@ -0,0 +1,143 @@
...
+    conn.log_expect(options["--command-prompt"], int(options["--login-timeout"]))
+    result = fence_action(conn, options, set_power_status, get_power_status, get_power_status)
+    ##
+    ## Logout from system
+    ##
+    ## In some special unspecified cases it is possible that
+    ## connection will be closed before we run close(). This is not
+    ## a problem because everything is checked before.
+    ######
+    try:
```
Comment (oalbrigt) [inline agents/cdu/fence_cdu.py:132]: You should move this to a function, and add it as an exit handler:
https://github.com/ClusterLabs/fence-agents/blob/8d746be92f191aa289f13a3703031c122a5e6cf3/agents/vmware_rest/fence_vmware_rest.py#L222

File: fence-agents.spec.in [hunk 3]
```diff
@@ -41,6 +41,7 @@ fence-agents-aws \\
 fence-agents-azure-arm \\
 fence-agents-bladecenter \\
 fence-agents-brocade \\
+fence-agents-cdu \\
```
Comment (oalbrigt) [inline fence-agents.spec.in:44]: You have to add %package/%description/%files sections for it as well.

File: agents/cdu/fence_cdu.py [hunk 4]
```diff
@@ -0,0 +1,143 @@
+#!/usr/bin/python3
```
Comment (oalbrigt) [inline agents/cdu/fence_cdu.py:1]: This needs to be `#!@PYTHON@ -tt` to get set based on detected or specified value during configure/build.

File: agents/cdu/fence_cdu.py [hunk 5]
```diff
@@ -0,0 +1,143 @@
...
+## +---------------------------------------------+
+## Sentry Switched CDU   6a
+## Sentry Switched CDU   7.1c <trenn@suse.de>
+##
+## @note: ssh has not being tested
+##
+#####
+
+import sys, re, pexpect, atexit
+sys.path.append("/usr/share/fence")
```
Comment (oalbrigt) [inline agents/cdu/fence_cdu.py:36]: This also needs to be changed to `@FENCEAGENTSLIBDIR@` to get replaced with detected/specified path when you build the project.

**Review by oalbrigt — CHANGES_REQUESTED** [review]

File: agents/cdu/fence_cdu.py [hunk 6]
```diff
@@ -0,0 +1,137 @@
...
+            fail(EC_CONNECTION_LOST)
+        except pexpect.TIMEOUT:
+            fail(EC_TIMED_OUT)
+
+def disconnect(conn):
+    conn.sendline("LOGOUT")
+    conn.close()
+
+def main():
+    device_opt = [  "action", "ipaddr", "login", "secure", "port",
```
Comment (oalbrigt) [inline agents/cdu/fence_cdu.py:90]: You can also remove action and ipport here.

File: agents/cdu/fence_cdu.py (hunk 1 above)
Comment (watologo1) [reply agents/cdu/fence_cdu.py]: I removed most of it, including all ssh related options.
the old default -1 blowfish option showed that this won't work nowadays anymore anyway and it was mentioned untested at these times already.

**Comment by watologo1** [comment]
Should be better now... ?
Thanks for your quick review(s)!

**Comment by oalbrigt** [comment]
ok to test

**Comment by oalbrigt** [comment]
Looking good now. Just a bit too much cleanup, so just readd "secure" in device_opt and run `make xml-check` to see that it only complains about difference in description (and then run `make xml-upload`.

**Comment by watologo1** [comment]
Next round:
Added secure, new xml test file, due to previous description modifs (removed ssh comments) and also added manpage to the file section this time.

**Comment by oalbrigt** [comment]
Nice.

I'll do some testing to find what needs changing to pass that delay-check :)

**Review by oalbrigt — APPROVED** [review]

**Comment by oalbrigt** [comment]
You need to add "passwd" to device_opt (so you'll have to run `make xml-upload` again after doing so).

You can also add the "no_login", "no_password" combination if the devices support using SSH keys to login.

**Comment by watologo1** [comment]
Am Freitag, 25. Juni 2021, 09:37:04 CEST schrieb Oyvind Albrigtsen:
> You need to add "passwd" to device_opt (so you'll have to run `make
> xml-upload` again after doing so).

Ah, yes. No password option anymore.
 
> You can also add the "no_login", "no_password" combination if the devices
> support using SSH keys to login.

I realized that the "secure" option implies:
--ssh, --identify-file and some more.

I tried a bit, but I could not get ssh connection working (also not a native
one using ssh executable).

First I run into:
no matching key exchange method found. Their offer: diffie-hellman-group1-sha1

then after adding:
-oKexAlgorithms=+diffie-hellman-group1-sha1
I run into:
no matching host key type found. Their offer: ssh-dss

I went the Sentry configs up and down, but I could not find anything related 
to ssh keys, neither in the ssh, nor in the user or whatever options.

So I would leave the "secure" option out.
It can be added later if I or someone else finds out more...

Sending another version with "passwd" added...

Thanks,

  Thomas

**Comment by oalbrigt** [comment]
Sounds good to me.

**Comment by oalbrigt** [comment]
LGTM. Thanks.

**Comment by watologo1** [comment]
> LGTM. Thanks.

Thank you. This was a nice and efficient review/work. It's very much appreciated!

**Pull request merged**